### ISSUE TITLE
IDE0059 incorrectly raise with async inline method, fix would result in CS0165 

### ISSUE THREAD

**Issue by datvm** [dotnet/roslyn#74512]
**Version Used**: 17.10.4.21802

**Steps to Reproduce**:

1. Enter the following code:

```cs
class Foo
{

    public async Task DoFooAsync()
    {
        MyClass c = null!; // IDE0059 here

        await GetMyClassAsync();

        Console.WriteLine(c.Value + 1);

        async Task GetMyClassAsync()
        {
            c = await Task.FromResult(new MyClass());
        }
    }

}

class MyClass
{
    public int Value { get; set; }
}
```

2. `IDE0059` is raised. Without `= null!`, `CS0165` would be raised instead.
3. Use "Remove redundant assignment" option, the code become invalid:

```cs
    public async Task DoFooAsync()
    {
        await GetMyClassAsync();

        MyClass c;
        Console.WriteLine(c.Value + 1);
        // ...
```

Interestingly if `GetMyClassAsync` is not `async`, this issue doesn't happen and `IDE0059` is not raised at all.

**Expected Behavior**:

`IDE0059` should not show up.

**Actual Behavior**:

Invalid `IDE0059` show up.
